protocol/mqtt_paho/v2/protocol.go

```diff
@@ -0,0 +1,134 @@
+package mqtt_paho
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sync"
+
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/protocol"
+	"github.com/eclipse/paho.golang/paho"
+
+	cecontext "github.com/cloudevents/sdk-go/v2/context"
+)
+
+type Protocol struct {
+	client         *paho.Client
+	connConfig     *paho.Connect
+	senderTopic    string
+	receiverTopics []string
+	qos            byte
+	retained       bool
+
+	// receiver
+	incoming chan *paho.Publish
+	// inOpen
+	openerMutex sync.Mutex
+}
+
+var (
+	_ protocol.Sender   = (*Protocol)(nil)
+	_ protocol.Opener   = (*Protocol)(nil)
+	_ protocol.Receiver = (*Protocol)(nil)
+	_ protocol.Closer   = (*Protocol)(nil)
+)
+
+func New(ctx context.Context, clientConfig *paho.ClientConfig, connConfig *paho.Connect, SenderTopic string,
+	ReceiverTopics []string, qos byte, retained bool,
+) (*Protocol, error) {
+	client := paho.NewClient(*clientConfig)
+	ca, err := client.Connect(ctx, connConfig)
+	if err != nil {
+		return nil, err
+	}
+	if ca.ReasonCode != 0 {
+		return nil, fmt.Errorf("failed to connect to %s : %d - %s", client.Conn.RemoteAddr(), ca.ReasonCode,
+			ca.Properties.ReasonString)
+	}
+
+	return &Protocol{
+		client:         client,
+		connConfig:     connConfig,
+		senderTopic:    SenderTopic,
+		receiverTopics: ReceiverTopics,
+		qos:            qos,
+		retained:       retained,
+		incoming:       make(chan *paho.Publish),
+		openerMutex:    sync.Mutex{},
+	}, nil
+}
+
+func (t *Protocol) Send(ctx context.Context, m binding.Message, transformers ...binding.Transformer) error {
+	var err error
+	defer m.Finish(err)
+
+	topic := cecontext.TopicFrom(ctx)
+	if topic == "" {
+		topic = t.senderTopic
+	}
+
+	msg := &paho.Publish{
+		QoS:    t.qos,
+		Retain: t.retained,
+		Topic:  topic,
```

Comment:
can we allow to add properties here?

Reply:
We can add properties in the event extensions and it will be propagated to the `paho.Pulish.Properties`.